src/main/java/com/cloudbees/plugins/credentials/ContextInPath.java

```diff
@@ -0,0 +1,46 @@
+package com.cloudbees.plugins.credentials;
+
+import hudson.model.ModelObject;
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import javax.servlet.ServletException;
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.stapler.AnnotationHandler;
+import org.kohsuke.stapler.InjectedParameter;
+import org.kohsuke.stapler.StaplerRequest;
+
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Indicates that this parameter is injected by evaluating
+ * {@link StaplerRequest#getAncestors()} and searching for a credentials context with the parameter type.
+ * You can enhance the lookup by ensuring that there are query parameters of {@code $provider} and {@code $token}
+ * that correspond to the context's {@link CredentialsSelectHelper.ContextResolver} FQCN and
+ * {@link CredentialsSelectHelper.ContextResolver#getToken(ModelObject)} respectively.
+ *
+ * @see CredentialsDescriptor#getCheckMethod(String)
+ * @since 2.1.5
+ */
+@Retention(RUNTIME)
+@Target(PARAMETER)
+@Documented
+@InjectedParameter(ContextInPath.HandlerImpl.class)
+public @interface ContextInPath {
```

Comment:
@jglick you may be interested in this as a less hacky approach to https://github.com/jenkinsci/workflow-step-api-plugin/blob/aabbdca1c3c1f1aacf346b01d95d3be467547195/src/main/java/org/jenkinsci/plugins/workflow/util/StaplerReferer.java#L43

I wonder could we introduce some extension points and make a more generic solution in core?